private/buf/cmd/buf/command/beta/convert/convert_test.go

```diff
@@ -0,0 +1,192 @@
+			"--type=buf.Foo",
+			"--payload=testdata/bin_json/payload.bin",
+			"testdata/bin_json/buf.proto",
+		)
+	})
+
+	t.Run("json-to-bin-file-proto", func(t *testing.T) {
+		testRun(
+			t,
+			nil,
+			"testdata/bin_json/payload.bin",
+			"--type=buf.Foo",
+			"--payload=testdata/bin_json/payload.json",
+			"testdata/bin_json/buf.proto",
+		)
+	})
+
+	t.Run("stdin-json-to-bin-proto", func(t *testing.T) {
+		testRun(
+			t,
+			strings.NewReader(`{"one":"55"}`),
+			"testdata/bin_json/payload.bin",
+			"--type=buf.Foo",
+			"--payload",
+			"-#format=json",
+			"testdata/bin_json/buf.proto",
+		)
+	})
+
+	t.Run("stdin-bin-to-json-proto", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/payload.bin")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.json",
+			"--type=buf.Foo",
+			"--payload",
+			"-#format=bin",
+			"testdata/bin_json/buf.proto",
+		)
+	})
+
+	t.Run("stdin-json-to-json-proto", func(t *testing.T) {
+		testRun(
+			t,
+			strings.NewReader(`{"one":"55"}`),
+			"testdata/bin_json/payload.json",
+			"--type=buf.Foo",
+			"testdata/bin_json/buf.proto",
+			"--payload",
+			"-#format=json",
+			"-o",
+			"-#format=json",
+		)
+	})
+
+	t.Run("stdin-input-to-json-image", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/image.bin")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.json",
+			"--type=buf.Foo",
+			"-",
+			"--payload=testdata/bin_json/payload.bin",
+			"-o",
+			"-#format=json",
+		)
+	})
+
+	t.Run("stdin-json-to-json-image", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/payload.bin")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.json",
+			"--type=buf.Foo",
+			"testdata/bin_json/image.bin",
+			"--payload",
+			"-#format=bin",
+			"-o",
+			"-#format=json",
+		)
+	})
+
+	t.Run("stdin-bin-payload-to-json-with-image", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/payload.bin")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.json",
+			"--type=buf.Foo",
+			"testdata/bin_json/image.bin",
+			"-o",
+			"-#format=json",
+		)
+	})
+
+	t.Run("stdin-json-payload-to-bin-with-image", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/payload.json")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.bin",
+			"--type=buf.Foo",
+			"testdata/bin_json/image.bin",
+			"--payload",
+			"-#format=json",
+			"-o",
+			"-#format=bin",
+		)
+	})
+
+	t.Run("stdin-image-json-to-bin", func(t *testing.T) {
+		file, err := os.Open("testdata/bin_json/image.json")
+		require.NoError(t, err)
+		testRun(
+			t,
+			file,
+			"testdata/bin_json/payload.bin",
+			"--type=buf.Foo",
+			"-#format=json",
+			"--payload=testdata/bin_json/payload.json",
+			"-o",
+			"-#format=bin",
+		)
+	})
+}
+
+func testRun(t *testing.T, stdin io.Reader, want string, args ...string) {
```

Comment:
probably better to give a more specific name to it?

Reply:
> I mean the &f.Input here, sorry if that is confused

This actually is the "Input" of proto/image/bsr fqn so I think it works